File: 0094. Binary Tree Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        stack, ret = [], []

        def DFS(node):
            while node:
                stack.append(node)
                node = node.left
            return
        
        DFS(root)

        while stack:
            node = stack.pop()
            ret.append(node.val)
            DFS(node.right)

        return ret

File: 0094. Binary Tree Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        std::vector<TreeNode*> stack;
        std::vector<int> ret;

        DFS(stack, root);

        while (!stack.empty()) {
            TreeNode* node = stack.back();
            stack.pop_back();
            ret.push_back(node->val);
            DFS(stack, node->right);
        }

        return ret; 
    }
private:
    void DFS(vector<TreeNode*> &stack, TreeNode* &node) {
        while(node) {
            stack.push_back(node);
            node = node->left;
        }
    }
};
